Libplanet/Net/Messages/Message.cs

```diff
@@ -94,26 +97,67 @@ internal enum MessageType : byte
             /// Contains the chain status of the peer at the moment.
             /// </summary>
             ChainStatus = 0x21,
+
+            /// <summary>
+            /// A reply to any messages with different <see cref="AppProtocolVersion"/>.
+            /// Contains the expected and actual <see cref="AppProtocolVersion"/>
+            /// value of the message.
+            /// </summary>
+            DifferentVersion = 0x30,
         }
 
         public byte[] Identity { get; set; }
 
+        public AppProtocolVersion Version { get; set; }
+
         public Peer Remote { get; set; }
 
         protected abstract MessageType Type { get; }
 
         protected abstract IEnumerable<NetMQFrame> DataFrames { get; }
 
-        public static Message Parse(NetMQMessage raw, bool reply)
+        public static Message Parse(
+            NetMQMessage raw,
+            bool reply,
+            AppProtocolVersion localVersion,
+            IImmutableSet<PublicKey> trustedAppProtocolVersionSigners,
+            DifferentAppProtocolVersionEncountered differentAppProtocolVersionEncountered)
         {
             if (raw.FrameCount == 0)
             {
                 throw new ArgumentException("Can't parse empty NetMQMessage.");
             }
 
-            // (reply == true)  [type, sign, peer, frames...]
-            // (reply == false) [identity, type, sign, peer, frames...]
-            int headerCount = reply ? 3 : 4;
+            // (reply == true)  [version, type, sign, peer, frames...]
+            // (reply == false) [identity, version, type, sign, peer, frames...]
+            int headerCount = reply ? CommonFrames : CommonFrames + 1;
+            var versionToken = raw[headerCount - 4].ConvertToString();
```

Comment:
IMO If we use a constant here, it will be easy to make mistakes. if changed `CommonFrames`, easily forgot about change const number. Can we change like that?

```suggestion
            var versionToken = raw[headerCount - CommonFrames].ConvertToString();
```
or something more semantic variable name.

Reply:
Then these lines:

https://github.com/planetarium/libplanet/blob/d46118902d7876c7ea20f25c7eedccf33677d9fa/Libplanet/Net/Messages/Message.cs#L161-L163

should be like this as well?

```
 var rawType = (MessageType)raw[headerCount - (CommonFrames - 1)].ConvertToInt32(); 
 var peer = raw[headerCount - (CommonFrames - 2)].ToByteArray(); 
 byte[] signature = raw[headerCount - (CommonFrames - 3)].ToByteArray(); 
```

Reply:
I think it seems OK. any opinion? @libplanet

Reply:
Or, it would be more readable if we slice the `raw` array after checking 'reply'.

```charp
byte[] remains  = (reply) ? raw.Skip(1).ToArray() : raw;

// using remains instead of raw
```

Reply:
I've added `MessageFrame` enum in [`e8eacb83f1d6f9a96d363f791b38752bc781884d`](https://github.com/planetarium/libplanet/pull/949/files#diff-4008fb0720717f5ed3e957339adebf4dR120).